src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,330 @@
+  const goto_programt::instructiont &instr,
+  const dstringt &identifier)
+{
+  // If not an assignment, return false
+  if(!instr.is_assign())
+  {
+    return false;
+  }
+  const auto &rhs=to_code_assign(instr.code).rhs();
+  return is_symbol_with_id(rhs, identifier) ||
+         is_typecast_with_id(rhs, identifier);
+}
+
+/*******************************************************************\
+
+Function: process_target
+
+  Inputs:
+    message_handler: Handles logging.
+    symbol_table: The table of known symbols.
+    max_nondet_array_length: Max length of arrays in new nondet objects.
+    instructions: The list of all instructions.
+    instr: The instruction to check.
+
+ Outputs: The next instruction to process, probably with this function.
+
+ Purpose: Given an iterator into a list of instructions, modify the list to
+          replace 'nondet' library functions with CBMC-native nondet
+          expressions, and return an iterator to the next instruction to check.
+
+\*******************************************************************/
+
+static goto_programt::targett process_target(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  const size_t max_nondet_array_length,
+  goto_programt &prog,
+  const goto_programt::targett &instr)
+{
+  // Check whether this is our nondet library method, and return if not
+  const auto instr_info=get_nondet_instruction_info(instr);
+  const auto next_instr=std::next(instr);
+  if(!instr_info.is_nondet_instruction)
+  {
+    return next_instr;
+  }
+
+  // Look at the next instruction, ensure that it is an assignment
+  assert(next_instr->is_assign());
+  // Get the name of the LHS of the assignment
+  const auto &next_instr_assign_lhs=to_code_assign(next_instr->code).lhs();
+  if(!(next_instr_assign_lhs.id()==ID_symbol &&
+       !next_instr_assign_lhs.has_operands()))
+  {
+    return next_instr;
+  }
+  const auto return_identifier=
+    to_symbol_expr(next_instr_assign_lhs).get_identifier();
+
+  auto &instructions=prog.instructions;
+  const auto end=std::end(instructions);
+
+  // Look for an instruction where this name is on the RHS of an assignment
+  const auto matching_assignment=std::find_if(
+    next_instr,
+    end,
+    [&return_identifier](const goto_programt::instructiont &instr)
+    {
+      return is_assignment_from(instr, return_identifier);
+    });
+
+  assert(matching_assignment!=end);
+
+  // Assume that the LHS of *this* assignment is the actual nondet variable
+  const auto &code_assign=to_code_assign(matching_assignment->code);
+  const auto nondet_var=code_assign.lhs();
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
+  // Generate nondet init code
+  code_blockt init_code;
```

Comment:
I think this should be done as follows:
1. remove here the nondet_*() calls and replace them by return_value = NONDET; for objects the classcast must be considered to get the right return type.
2. in a separate, language-independent goto transformation pass, e.g. remove_nondet(), elaborate the NONDET of complex types by appropriate initialisation code. This has the advantage that the nondet-initialisation code that is currently present in *_entry_point can go away and is uniformly handled by this new pass. Moreover, it's not language-dependent anymore and more easily maintainable.